pkg/store/bucket.go

```diff
@@ -2092,93 +2104,84 @@ func (r *bucketIndexReader) Close() error {
 			return errors.Wrap(err, "lookup label name")
 		}
-		lv, err := r.dec.LookupSymbol(lvo)
+		lv, err := r.dec.LookupSymbol(s.value)
 		if err != nil {
 			return errors.Wrap(err, "lookup label value")
 		}
-
 		*lbls = append(*lbls, labels.Label{Name: ln, Value: lv})
 	}
+	return nil
+}
 
-	// Read the chunks meta data.
-	k = d.Uvarint()
+// decodeSeriesForTime decodes a series entry from the given byte slice decoding only chunk metas that are within given min and max time.
+// If skipChunks is specified decodeSeriesForTime does not return any chunks, but only labels and only if at least single chunk is within time range.
+// decodeSeriesForTime returns false, when there are no series data for given time range.
+func decodeSeriesForTime(b []byte, lset *[]symbolizedLabel, chks *[]chunks.Meta, skipChunks bool, selectMint, selectMaxt int64) (ok bool, err error) {
+	*lset = (*lset)[:0]
 
-	if k == 0 {
-		return nil
+	if !skipChunks {
+		*chks = (*chks)[:0]
 	}
 
-	t0 := d.Varint64()
-	maxt := int64(d.Uvarint64()) + t0
-	ref0 := int64(d.Uvarint64())
+	d := encoding.Decbuf{B: b}
 
-	// No chunk in the required time range.
-	if t0 > req.MaxTime {
-		return nil
+	// Read labels without looking up symbols.
+	k := d.Uvarint()
+	for i := 0; i < k; i++ {
+		lno := uint32(d.Uvarint())
+		lvo := uint32(d.Uvarint())
+		*lset = append(*lset, symbolizedLabel{name: lno, value: lvo})
 	}
-
-	if req.MinTime <= maxt {
-		*chks = append(*chks, chunks.Meta{
-			Ref:     uint64(ref0),
-			MinTime: t0,
-			MaxTime: maxt,
-		})
-		// Get a valid chunk, return if it is a skip chunk request.
-		if req.SkipChunks {
-			return nil
-		}
+	// Read the chunks meta data.
+	k = d.Uvarint()
+	if k == 0 {
+		return false, d.Err()
 	}
-	t0 = maxt
 
-	for i := 1; i < k; i++ {
-		mint := int64(d.Uvarint64()) + t0
-		maxt := int64(d.Uvarint64()) + mint
-		ref0 += d.Varint64()
-		t0 = maxt
+	// First t0 is absolute, rest is just diff so different type is used (Uvarint64).
+	mint := d.Varint64()
+	maxt := int64(d.Uvarint64()) + mint
+	// Similar for first ref.
+	ref := int64(d.Uvarint64())
 
-		if maxt < req.MinTime {
-			continue
-		}
-		if mint > req.MaxTime {
-			break
+	for i := 0; i < k; i++ {
+		if i > 0 {
+			mint += int64(d.Uvarint64())
+			maxt = int64(d.Uvarint64()) + mint
+			ref += d.Varint64()
 		}
 
-		if d.Err() != nil {
-			return errors.Wrapf(d.Err(), "read meta for chunk %d", i)
+		if mint > selectMaxt {
+			break
 		}
 
-		*chks = append(*chks, chunks.Meta{
-			Ref:     uint64(ref0),
-			MinTime: mint,
-			MaxTime: maxt,
-		})
+		if maxt >= selectMint {
+			// Found a chunk.
+			if skipChunks {
+				// We are not interested in chunks and we know there is at least one, that's enough to return series.
+				return true, nil
+			}
 
-		if req.SkipChunks {
-			return nil
+			*chks = append(*chks, chunks.Meta{
+				Ref:     uint64(ref),
+				MinTime: mint,
+				MaxTime: maxt,
+			})
 		}
+
+		mint = maxt
 	}
-	return d.Err()
+
+	return len(*chks) > 0, d.Err()
```

Comment:
For `skipChunks` query, since we are not resetting the chunk slice, this always returns true even there is no matching chunk.
Please correct me if I was wrong here.

Reply:
you are indeed right, thanks for spotting :hugs:

Reply:
will fix